### ISSUE TITLE
💡Self hosted Cloudflare Tunnel server side

### ISSUE THREAD

**Issue by ShiinaRinne** [cloudflare/cloudflared#1175]
**Describe the feature you'd like**
In certain regions, using Cloudflare Tunnel may introduce significant latency. 
Therefore, I’m interested in deploying the backend on my own server, which is closer to the visitors, instead of accessing through Cloudflare’s servers.

**Describe alternatives you've considered**
I’ve tried [nps](https://github.com/ehang-io/nps) and also considered writing my own management tools to implement this feature with [frp](https://github.com/fatedier/frp). 
However, I ultimately want to do this because I prefer the way Cloudflare Tunnel works and aim to emulate it. 

**Additional context**
If Cloudflare Tunnel could support self-hosting of the backend, it would likely attract many users who are facing similar latency issues. This feature would enhance the flexibility and adaptability of Cloudflare Tunnel, making it an even more powerful tool for a wider range of applications.

I'm sorry that I have no experience with Go or related fields, so I’m not sure where to start implementing this feature after briefly reviewing the code

**Comment by ShiinaRinne** [comment]
> Hello, You can use Docker to create backend. You can also create frontend in Docker, too. Then expose your frontend via Cloudflare Tunnel.

Thank you for your reply~ 
But after I posted this issue, I did some more searching, and it seems that the server side of Cloudflare Tunnel is closed source.

What you’re suggesting seems to be deploying a backend service in the intranet on my own, and then making it public to the internet through Cloudflare Tunnel, but this is not what I’m looking for. 
Similar to the server and client in [FRP](https://github.com/fatedier/frp), Cloudflared is the client side, and my idea is to deploy the server side on my own server, then manage it directly from the Zero Trust panel, or generate a private site for configuration. 
I’m not sure if Cloudflare will support this method, but now I  think there’s almost no hope.

Although I have the idea of reverse engineering, and I have done some simple reverse engineering work before, but I don’t have much experience, and I can’t implement this kind of service just by analyzing the code of Cloudflared.

**Comment by jcsf** [comment]
Cloudflared heavily relies on multiple Cloudflare services, therefore it is not possible to self-host the backend, you wouldn't even have enough hardware power for that :wink:. Therefore, I will close this ticket since it is not possible to be done.

**Comment by thebalaa** [comment]
Check out https://github.com/hintjen/selfhosted-gateway @ShiinaRinne

**Comment by ShiinaRinne** [comment]
> Check out https://github.com/hintjen/selfhosted-gateway [@ShiinaRinne](https://github.com/ShiinaRinne)

I clicked into this issue from a year ago with excitement and saw your reply. I thought it was a reverse implementation of cloudflare tunnel, after checking the project, I found that I had already starred the project, and I never even looked at it carefully and used it.  
But what I want is actually a tool that is more in line with the cloudflare tunnel ecosystem, such as zerotier's moon and tailscale's derp (or GitHub Action runner?)  mentioned in https://github.com/cloudflare/cloudflared/issues/1413. This allows users to forward traffic through self-hosted nodes without having to go through cloudflare's official servers.

But cloudflare seems to have no further development plans at present. They think it is "not possible to be done", so this demand should never be solved.

Despite this, thank you very much for your reply~
